src/tree/registries/auth/IAuthHelper.ts

```diff
@@ -0,0 +1,21 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See LICENSE.md in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import * as request from 'request-promise-native';
+import { URL } from 'url';
+import { ICachedRegistryProvider } from '../ICachedRegistryProvider';
+import { IDockerCliCredentials } from '../RegistryTreeItemBase';
+
+export interface IAuthHelper {
+    addAuth(cachedProvider: ICachedRegistryProvider, options: request.RequestPromiseOptions, authContext?: IOAuthContext): Promise<void>;

```

Comment:
I'm not a fan of methods that modify incoming data.  Would it be better to return a new `request.RequestPromiseOptions` that's a (at least a shallow) copy the incoming `options` but with, for example, the auth properties set?  It's a pattern that request options are created once and then reused between requests; callers may not be expecting that object to be modified.

Reply:
For this we can return `AuthOptions` or whatever it's called. To change `addAuth` on tree nodes would require refactoring the other providers; we certainly can do that but we should consider that.